lib/irb/ruby-lex.rb

```diff
@@ -69,64 +69,38 @@ def configure_io(io)
           # Accept any single-line input for symbol aliases or commands that transform args
           next true if single_line_command?(code)
 
-          ltype, indent, continue, code_block_open = check_code_state(code)
-          if ltype or indent > 0 or continue or code_block_open
-            false
-          else
-            true
-          end
+          _tokens, _opens, terminated = check_code_state(code)
+          terminated
         end
       end
     end
     if @io.respond_to?(:dynamic_prompt)
       @io.dynamic_prompt do |lines|
         lines << '' if lines.empty?
-        result = []
         tokens = self.class.ripper_lex_without_warning(lines.map{ |l| l + "\n" }.join, context: @context)
-        code = String.new
-        partial_tokens = []
-        unprocessed_tokens = []
-        line_num_offset = 0
-        tokens.each do |t|
-          partial_tokens << t
-          unprocessed_tokens << t
-          if t.tok.include?("\n")
-            t_str = t.tok
-            t_str.each_line("\n") do |s|
-              code << s
-              next unless s.include?("\n")
-              ltype, indent, continue, code_block_open = check_state(code, partial_tokens)
-              result << @prompt.call(ltype, indent, continue || code_block_open, @line_no + line_num_offset)
-              line_num_offset += 1
-            end
-            unprocessed_tokens = []
-          else
-            code << t.tok
+        line_results = IRB::NestingParser.parse_by_line(tokens)
+        tokens_until_line = []
+        line_results.map.with_index do |(line_tokens, _prev_opens, next_opens, _min_depth), line_num_offset|
+          line_tokens.each do |token, _s|
+            tokens_until_line << token if token != tokens_until_line.last
```

Comment:
Why do we need this if condition? Looking at `process_continue`, having this token or not should make a difference. But even if I removed this line, all tests still pass.

Reply:
For this code
```ruby
if 1
  "aaa
  bbb
  ccc"
end
```
tokens on line 1 are: `["if", " ", "1", "\n"]`
tokens on line 2 are: `["  ", '"', "aaa\n  bbb\n  ccc"]`
tokens on line 3 are: `["aaa\n  bbb\n  ccc"]`
tokens on line 4 are: `["aaa\n  bbb\n  ccc", '"', "\n"]`
tokens on line 5 are: `["end"]`
It has duplication because token `on_tstring_content` can span multiline.

If I remove this condition, tokens_until_line will be:
`["if", " ", "1", "\n", "  ", "\"", "aaa\n  bbb\n  ccc", "aaa\n  bbb\n  ccc", "aaa\n  bbb\n  ccc", "\"", "\n", "end"]`
`"aaa\n  bbb\n  ccc"` is duplicated three times.

It will work without this condition because duplication does not affect the result of `process_continue.

Reply:
Given it's for optimisation purpose and currently can't be covered by tests, I think we can add a comment like:

```suggestion
            # avoid appending duplicated last tokens to keep the list small
            tokens_until_line << token if token != tokens_until_line.last
```

Reply:
Oh I was confused by this:
>  but It will work without this condition because duplication does not affect the result of process_continue.

That comment looks good 👍